crates/accelerate/src/remove_diagonal_gates_before_measure.rs

```diff
@@ -0,0 +1,105 @@
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+/// Run the RemoveDiagonalGatesBeforeMeasure pass on `dag`.
+/// Args:
+///     dag (DAGCircuit): the DAG to be optimized.
+/// Returns:
+///     DAGCircuit: the optimized DAG.
+#[pyfunction]
+#[pyo3(name = "remove_diagonal_gates_before_measure")]
+fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()> {
+    static DIAGONAL_1Q_GATES: [StandardGate; 8] = [
+        StandardGate::RZGate,
+        StandardGate::ZGate,
+        StandardGate::TGate,
+        StandardGate::SGate,
+        StandardGate::TdgGate,
+        StandardGate::SdgGate,
+        StandardGate::U1Gate,
+        StandardGate::PhaseGate,
+    ];
+    static DIAGONAL_2Q_GATES: [StandardGate; 7] = [
+        StandardGate::CZGate,
+        StandardGate::CRZGate,
+        StandardGate::CU1Gate,
+        StandardGate::RZZGate,
+        StandardGate::CPhaseGate,
+        StandardGate::CSGate,
+        StandardGate::CSdgGate,
+    ];
+
+    let mut nodes_to_remove = Vec::new();
+    for index in dag.op_nodes(true) {
+        let node = &dag.dag[index];
+        let NodeType::Operation(inst) = node else {panic!()};
+
+        if inst.op.name() == "measure" {
+            let predecessor = (dag.quantum_predecessors(index))
+                .next()
+                .expect("index is an operation node, so it must have a predecessor.");
+
+            match &dag.dag[predecessor] {
+                NodeType::Operation(pred_inst) => match pred_inst.standard_gate() {
+                    Some(gate) => {
+                        if DIAGONAL_1Q_GATES.contains(&gate) {
+                            nodes_to_remove.push(predecessor);
+                        } else if DIAGONAL_2Q_GATES.contains(&gate) {
+                            let successors = dag.quantum_successors(predecessor);
+                            let remove_s = successors
+                                .map(|s| {
+                                    let node_s = &dag.dag[s];
+                                    if let NodeType::Operation(inst_s) = node_s {
+                                        inst_s.op.name() == "measure"
+                                    } else {
+                                        false
+                                    }
+                                })
+                                .all(|ok_to_remove| ok_to_remove);
+                            if remove_s {
+                                nodes_to_remove.push(predecessor);
+                            }
+                        }
+                    }
+                    None => {
+                        continue;
+                    }
+                },
+                _ => {
+                    continue;
+                }
+            }
+        }
+    }
+
+    let set_nodes_to_remove: HashSet<NodeIndex> = nodes_to_remove.into_iter().collect();
+    for node_to_remove in set_nodes_to_remove {
+        dag.remove_op_node(node_to_remove)
+    }
```

Comment:
You can avoid the deduplication with a hash set here by just check doing a quick lookup on the graph before calling remove. Something like:

```suggestion
    for node_to_remove in nodes_to_remove {
        if dag.dag.node_weight(node_to_remove).is_some() {
            dag.remove_op_node(node_to_remove);
        }
    }
```

Reply:
| n_qubits | depth | time is sec  | 
|----------|--------|--------------|
Actually, this change makes the code slower, so I didn't change the current code with the HashSet.

| n_qubits| depth | time in sec |
|----------|--------|-------------|
| 5 | 1024 | 0.01 |  
| 14 | 1024 | 0.05 |  
| 20 | 1024 | 0.08 | 
| 50 | 1024 |  0.53 |  
| 100 | 1024 | 2.85 |